src/tools/miri/src/shims/foreign_items.rs

```diff
@@ -346,11 +345,8 @@ pub trait EvalContextExt<'mir, 'tcx: 'mir>: crate::MiriInterpCxExt<'mir, 'tcx> {
 
         match allocator_kind {
             AllocatorKind::Global => {
-                let (body, instance) = this
-                    .lookup_exported_symbol(symbol)?
-                    .expect("symbol should be present if there is a global allocator");
-
-                Ok(EmulateByNameResult::MirBody(body, instance))
+                // `__rust_*` is defined by `#[global_allocator]` if `#[global_allocator]` is used
+                return Ok(EmulateByNameResult::NotSupported);
```

Comment:
I don't understand this comment. It seems unrelated both to the preceding match and the following Unsupported error. What is going on here? The entire set of Miri allocator changes is somewhat mysterious.

Reply:
Returning `NotSupported` will be interpreted as there existing no shim. If there is no shim it will fall back to looking for an `#[no_mangle]` symbol with the right name. Exactly what is necessary here. Only when `#[global_allocator]` is not used, should `__rust_*` be redirected to the default `__rdl_*` implementations in libstd.

Reply:
Oh you want to hit the fallback path in emulate_foreign_item. This needs a *ton* of comments explaining what is going on.^^ Both here and also below where you moved the `check_shim` calls around.

Reply:
This better?